src/humanloop/otel/processor.py

```diff
@@ -0,0 +1,173 @@
+import json
+import logging
+from collections import defaultdict
+
+import parse
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SimpleSpanProcessor, SpanExporter
+
+from humanloop.otel.constants import HL_FILE_OT_KEY, HL_LOG_OT_KEY
+from humanloop.otel.helpers import (
+    is_humanloop_span,
+    is_llm_provider_call,
+    read_from_opentelemetry_span,
+    write_to_opentelemetry_span,
+)
+
+
+class HumanloopSpanProcessor(SimpleSpanProcessor):
+    """Merge information from Instrumentors used by Humanloop SDK into the
+    Spans that will be exported to Humanloop.
+    """
+
+    def __init__(self, exporter: SpanExporter) -> None:
+        super().__init__(exporter)
+        self._spans = dict()
+        self._children = defaultdict(list)
+
+    # TODO: Could override on_start and process Flow spans ahead of time
+    # and PATCH the created Logs in on_end. A special type of ReadableSpan could be
+    # used for this
+
+    def on_end(self, span: ReadableSpan) -> None:
+        if is_humanloop_span(span=span):
+            _process_humanloop_span(span, self._children[span.context.span_id])
+            del self._children[span.context.span_id]
+            self.span_exporter.export([span])
+        else:
+            if span.parent is not None and _is_instrumentor_span(span):
+                self._children[span.parent.span_id].append(span)
+
+
+def _is_instrumentor_span(span: ReadableSpan) -> bool:
+    # TODO: Extend in the future as needed. Spans not coming from
+    # Instrumentors of interest should be dropped
+    return is_llm_provider_call(span=span)
+
+
+def _process_humanloop_span(span: ReadableSpan, children_spans: list[ReadableSpan]):
+    hl_file = read_from_opentelemetry_span(span, key=HL_FILE_OT_KEY)
+
+    if "prompt" in hl_file:
+        _process_prompt(prompt_span=span, children_spans=children_spans)
+        return
+    elif "tool" in hl_file:
+        _process_tool(tool_span=span, children_spans=children_spans)
+        return
+    elif "flow" in hl_file:
+        _process_flow(flow_span=span, children_spans=children_spans)
+        return
+    else:
+        logging.error("Invalid span type")
+
+
+def _process_prompt(prompt_span: ReadableSpan, children_spans: list[ReadableSpan]):
+    if len(children_spans) == 0:
+        return
+    child_span = children_spans[0]
+    assert is_llm_provider_call(child_span)
+    _enrich_prompt_span_file(prompt_span, child_span)
+    _enrich_prompt_span_log(prompt_span, child_span)
+
+
+def _process_tool(tool_span: ReadableSpan, children_spans: list[ReadableSpan]):
+    # TODO: Use children_spans in the future
```

Comment:
Expand on this todo - is it a blocker?

Reply:
No - children_spans is a placeholder for Spans created by Instrumentors that could enrich the HL Tool Span e.g. intercepting a call to Pinecone vector DB and adding the pinecone library version to the Tool File's attributes. 

I would leave those Instrumentors as future scope, since they're not critical for the DX, unlike the LLM provider Instrumentors that help us build most of the Prompt kernel